src/v/storage/mvlog/versioned_log.cc

```diff
@@ -0,0 +1,171 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "storage/mvlog/versioned_log.h"
+
+#include "container/fragmented_vector.h"
+#include "model/offset_interval.h"
+#include "storage/mvlog/file.h"
+#include "storage/mvlog/logger.h"
+#include "storage/mvlog/readable_segment.h"
+#include "storage/mvlog/segment_appender.h"
+
+#include <seastar/core/circular_buffer.hh>
+#include <seastar/core/io_priority_class.hh>
+#include <seastar/core/lowres_clock.hh>
+
+#include <chrono>
+
+using namespace std::literals::chrono_literals;
+
+namespace storage::experimental::mvlog {
+
+active_segment::active_segment(
+  std::unique_ptr<file> f, model::offset o, segment_id id, size_t target_size)
+  : segment_file(std::move(f))
+  , appender(std::make_unique<segment_appender>(segment_file.get()))
+  , readable_seg(std::make_unique<readable_segment>(segment_file.get()))
+  , construct_time(ss::lowres_clock::now())
```

Comment:
would construction time be attached to the file instead?

Reply:
Do you mean whether construction time should be something tracked in the `mvlog::file`? I think probably not -- the construction time is only really useful for now in determining if we should apply segment.ms

Or are you asking whether it should be attached to something other than the file construction (e.g. first data write)? I think that wouldn't be a bad idea -- maybe it'd be important to ensure no "empty" segments e.g. if a segment contains only truncate entries maybe we wouldn't want it being rolled? Though I'm not sold that such empty segments are a problem

Reply:
I was thinking that `ctime` is a normal posix file property, so associating it with the actual file made some sense. but maybe it's just a naming collision? we probably want to track our own creation timestamp if that timestamp is going to be used for retention purposes.

Reply:
Ah gotcha. Yeah it probably makes sense for retention. Will ponder a bit on how to resolve the naming conflict. I agree it might be confusing down the line